### ISSUE TITLE
Custom Enemy HUD not working - V0.202.14 (Hearth&Home)

### ISSUE THREAD

**Issue by RazorsEdge88** [matias-kovero/BetterUI#10]
Custom Enemy HUD from this mod is no longer working as seen in this picture

![image](https://user-images.githubusercontent.com/71830790/133688801-3a34127f-ea2f-4e56-93d9-838c4525bc29.png)

disabling this options does nothing and the default enemy hud cant be shown as long as this mod is loaded

`## Toggle the use of custom enemy hud`
`# Setting type: Boolean`
`# Default value: true`
`customEnemyHud = false`

**Comment by Dalayeth** [comment]
This is due to the removal of the "m_healthRoot" in enemy GUI's. The way the default game displays health has been changed a bit for mounts.

Lines 30-56 in "GameClasses/EnemyHud.cs" should probably be replaced with the new code outright.

Then the line
`RectTransform hpRoot = (hudData.m_healthRoot.transform as RectTransform);`
should be replaced with
`RectTransform hpRoot = (hudData.m_gui.transform.Find("Health") as RectTransform);`

Then there is a minor change to the GetSlot() function in Cooking Stations. The line
`cookingStation.GetSlot(i, out string text, out float num);`
Just needs to change to
`cookingStation.GetSlot(i, out string text, out float num, out CookingStation.Status status);`

That is how I got it working. I'll include a DLL that I made with these changes that works for me.
[BetterUI.zip](https://github.com/matias-kovero/BetterUI/files/7185993/BetterUI.zip)

**Comment by RazorsEdge88** [comment]
![image](https://user-images.githubusercontent.com/71830790/133843332-22577633-1868-43ed-a71b-80fed55406e1.png)

it works, thank you

**Comment by matias-kovero** [comment]
> This is due to the removal of the "m_healthRoot" in enemy GUI's. The way the default game displays health has been changed a bit for mounts.
> 
> Lines 30-56 in "GameClasses/EnemyHud.cs" should probably be replaced with the new code outright.
> 
> Then the line
> `RectTransform hpRoot = (hudData.m_healthRoot.transform as RectTransform);`
> should be replaced with
> `RectTransform hpRoot = (hudData.m_gui.transform.Find("Health") as RectTransform);`
> 
> Then there is a minor change to the GetSlot() function in Cooking Stations. The line
> `cookingStation.GetSlot(i, out string text, out float num);`
> Just needs to change to
> `cookingStation.GetSlot(i, out string text, out float num, out CookingStation.Status status);`

I have not updated the code in a while. If you are keen on creating an PR with the needed updates I'll review and publish it.
This mod might need some caring as the new update most likely broke a bunch of stuff(?).

**Comment by RazorsEdge88** [comment]
im testing it and trying to find any bugs, the only ones i found so far are:

-mob healthbars show the correct amount of in numbers (150/150HP for example) but that value doesnt refresh until the mobs gets out of range for the bar to dissapear from your UI... the red bar works, tho

-remaining time for buffs is always shown as 10m remaining, even when theres no food in that slot
![image](https://user-images.githubusercontent.com/71830790/133884895-2afca77d-ec5f-4211-a906-f3c2805cebd4.png)


currently using it with a bunch of others mods and found no conflicts so far, seems like the things that are broken are kinda minor for now

**Comment by matias-kovero** [comment]
As this was my 1st mod I created, it has some bad practices. 
Ex. Just copied to original code and not just returning to allow the original code to run.

This might be the issue now - as it will run my copied code from pre-patched Valheim.
So if the update updated some parts of the old code - it will break.

However, just checked what are the current options for UI mods, and it seems there is an good option [Project Auga](https://github.com/RandyKnapp/Auga)
It is made by Randy whom I chatted a few times and he is active with updates on his mods. 
That project will more likely make this UI mod obsolete.

**Comment by RazorsEdge88** [comment]
Sad to see this project die tho, will use this one with the correction dalayeth made and wait till RanyKnapp give us and update on their repository

**Comment by Mickroz** [comment]
ah, too bad, i really liked this project, because of the color scheme for items and such :( and custom layout.
Any way @RandyKnapp could merge these two projects?

**Comment by thedefside** [comment]
The problem with Auga is that any mod that makes changes to the default UI won't work with Auga unless the mod dev adds specific support for both UIs. I only know of 2 mods that currently have that.

**Comment by RandyKnapp** [comment]
@matias-kovero Hit me up on Discord (or here) and we'll see if we can collaborate and bring some BetterUI features to Auga

**Comment by nandryshak** [comment]
@Dalayeth Can you share all the code changes you made? I tried making them myself, but it doesn't work exactly like your DLL.

edit: Actually I think I got it. Thanks!

**Comment by nandryshak** [comment]
@RazorsEdge88 

I've fixed the health bars updating, will check out the food timers next. Any other bugs you noticed?

**Comment by IrcDirk** [comment]
@nandryshak can u share your work? Maybe fork + updates?

**Comment by nandryshak** [comment]
@IrcDirk Yeah definitely! I will open a PR later today. I looked into the food timers a little bit but haven't found the problem yet.

**Comment by baloche44** [comment]
@nandryshak not really a bug, but cooking timers are missing on the oven, if u can take a look at that <3

edit : my bad, u just post 3 minutes ago about it, sorry for that

**Comment by nandryshak** [comment]
@baloche44 Actually I think that's a separate issue, I was talking about the food times on the food bar (time remaining until the food that you've eaten expires)

**Comment by luxzg** [comment]
@nandryshak could you compile a new dll as a stop gap while we wait for full release? You obviously have it running in your game with fixes. It would mean a lot. @Dalayeth was kind enough to post one of fixed dlls, but there were other fixes after it as well in your (@nandryshak ) PRs. There is also a fork from @phtnk but no .dll either. Either of you can just attach it here as reply. Thank you in advance!

**Comment by nandryshak** [comment]
@luxzg Here you go! https://github.com/nandryshak/BetterUI/releases/tag/2021-10-20.1

Still a lot of things missing. For enemy levels, make sure you set it to stars only (which is not the default iirc), other options are broken.

**Comment by luxzg** [comment]
This is great, thank you! Jump works, food timer works, enemies at least have some HUD, it's better than we had before, I'm happy :)

If I may suggest, if @matias-kovero won't be maintaining this mod anymore, maybe he can allow the rest of you to publish it with fixes.

**Comment by thedefside** [comment]
@nandryshak I would like to contribute some changes, but I am unable to build your branch. I have all the dependencies resolved, but I am getting errors
```
'RectTransform' does not contain a definition for 'localEulerAngles'
'Camera' does not contain a definition for 'ViewportToScreenPoint'
The property or indexer 'RectTransform.offsetMin' cannot be used in this context because it lacks the get accessor
```

Any ideas what I'm doing wrong?

**Comment by nandryshak** [comment]
@thedefside Follow the instructions in the README under "Developing environment"

**Comment by filikun** [comment]
> @luxzg Here you go! https://github.com/nandryshak/BetterUI/releases/tag/2021-10-20.1
> 
> Still a lot of things missing. For enemy levels, make sure you set it to stars only (which is not the default iirc), other options are broken.

This solved a lot! Thanks! Any chance this will be pushed for release?

**Comment by thedefside** [comment]
> > @luxzg Here you go! https://github.com/nandryshak/BetterUI/releases/tag/2021-10-20.1
> > Still a lot of things missing. For enemy levels, make sure you set it to stars only (which is not the default iirc), other options are broken.
> 
> This solved a lot! Thanks! Any chance this will be pushed for release?

You can check out my version, which includes the fixes listed here plus more and new features

https://valheim.thunderstore.io/package/thedefside/BetterUI_Reforged/
